src/rgw/rgw_dedup.cc

```diff
@@ -0,0 +1,2363 @@
+    decode(ctl.started, bl);
+    decode(ctl.dedup_exec, bl);
+    decode(ctl.shutdown_req, bl);
+    decode(ctl.shutdown_done, bl);
+    decode(ctl.local_pause_req, bl);
+    decode(ctl.local_paused, bl);
+    decode(ctl.remote_abort_req, bl);
+    decode(ctl.remote_aborted, bl);
+    decode(ctl.remote_pause_req, bl);
+    decode(ctl.remote_paused, bl);
+    decode(ctl.remote_restart_req, bl);
+    DECODE_FINISH(bl);
+  }
+
+  //---------------------------------------------------------------------------
+  std::ostream& operator<<(std::ostream &out, const control_t &ctl)
+  {
+    out << ctl.dedup_type;
+    if (ctl.started) {
+      out << "::started";
+    }
+    if (ctl.dedup_exec) {
+      out << "::dedup_exec";
+    }
+    if (ctl.shutdown_req) {
+      out << "::shutdown_req";
+    }
+    if (ctl.shutdown_done) {
+      out << "::shutdown_done";
+    }
+    if (ctl.local_pause_req) {
+      out << "::local_pause_req";
+    }
+    if (ctl.local_paused) {
+      out << "::local_paused";
+    }
+    if (ctl.remote_abort_req) {
+      out << "::remote_abort_req";
+    }
+    if (ctl.remote_aborted) {
+      out << "::remote_aborted";
+    }
+    if (ctl.remote_pause_req) {
+      out << "::remote_pause_req";
+    }
+    if (ctl.remote_paused) {
+      out << "::remote_paused";
+    }
+    if (ctl.remote_restart_req) {
+      out << "::remote_restart_req";
+    }
+
+    return out;
+  }
+
+  //===========================================================================
+  // rgw::dedup::Background
+  //===========================================================================
+  //---------------------------------------------------------------------------
+  int Background::init_rados_access_handles()
+  {
+    store = dynamic_cast<rgw::sal::RadosStore*>(driver);
+    if (!store) {
+      ldpp_dout(dpp, 0) << "ERR: failed dynamic_cast to RadosStore" << dendl;
+      // this is the return code used in rgw_bucket.cc
+      return -ENOTSUP;
+    }
+
+    rados = store->getRados();
+    rados_handle = rados->get_rados_handle();
+
+    return init_dedup_pool_ioctx(rados, dpp, d_dedup_cluster_ioctx);
+  }
+
+  //---------------------------------------------------------------------------
+  Background::Background(rgw::sal::Driver* _driver, CephContext* _cct) :
+    driver(_driver),
+    dp(_cct, dout_subsys, "dedup background: "),
+    dpp(&dp),
+    cct(_cct),
+    d_cluster(dpp, cct, driver),
+    d_watcher_ctx(this)
+  {
+#if 0
+    // TBD: use config-file
+    // temp hack to allow processing small files
+    d_min_obj_size_for_dedup = 1024;
+#endif
+
+    if (init_rados_access_handles() != 0) {
+      throw std::runtime_error("Failed init_dedup_pool_ioctx()");
+    }
+    d_heart_beat_last_update = ceph_clock_now();
+    d_heart_beat_max_elapsed_sec = 3;
+  }
+
+  //---------------------------------------------------------------------------
+  Background::~Background()
+  {
+    // TBD:
+    // Do we need to stop the runner thread first ???
```

Comment:
i think that if you start the thread in a function (not ctor), it should also be stopped in  a function and not dtor